File: venefica-site/application/libraries/message_service.php
<?php if ( ! defined('BASEPATH')) exit('No direct script access allowed');

/**
 * Comment and message service
 *
 * @author gyuszi
 */
class Message_service {
    
    public function __construct() {
        log_message(DEBUG, "Initializing Message_service");
    }
    
    //**************
    //* commenting *
    //**************
    
    /**
     * 
     * @param long $adId
     * @param Comment_model $comment
     * @return long comment id
     * @throws Exception
     */
    public function addCommentToAd($adId, $comment) {
        try {
            $messageService = new SoapClient(MESSAGE_SERVICE_WSDL, getSoapOptions(loadToken()));
            $result = $messageService->addCommentToAd(array(
                "adId" => $adId,
                "comment" => $comment
            ));
            return $result->commentId;
        } catch ( Exception $ex ) {
            log_message(ERROR, 'Adding comment to ad (adId: ' . $adId . ') failed! '.$ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    /**
     * 
     * @param Comment_model $comment
     * @throws Exception
     */
    public function updateComment($comment) {
        try {
            $messageService = new SoapClient(MESSAGE_SERVICE_WSDL, getSoapOptions(loadToken()));
            $messageService->addCommentToAd(array("comment" => $comment));
        } catch ( Exception $ex ) {
            log_message(ERROR, 'Update comment failed! '.$ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    /**
     * 
     * @param long $adId
     * @param long $lastCommentId
     * @param int $numComments
     * @return array of Comment_model
     * @throws Exception
     */
    public function getCommentsByAd($adId, $lastCommentId, $numComments) {
        try {
            $messageService = new SoapClient(MESSAGE_SERVICE_WSDL, getSoapOptions(loadToken()));
            $result = $messageService->getCommentsByAd(array(
                "adId" => $adId,
                "lastCommentId" => $lastCommentId,
                "numComments" => $numComments
            ));
            
            $comments = array();
            if ( hasField($result, 'comment') && $result->comment ) {
                $comments = Comment_model::convertComments($result->comment);
            }
            return $comments;
        } catch ( Exception $ex ) {
            log_message(ERROR, 'Comments request failed! '.$ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    //*************
    //* messaging *
    //*************
    
    /**
     * 
     * @return array of Message_model
     * @throws Exception
     */
    public function getLastMessagePerRequest() {
        try {
            $messageService = new SoapClient(MESSAGE_SERVICE_WSDL, getSoapOptions(loadToken()));
            $result = $messageService->getLastMessagePerRequest();
            
            $messages = array();
            if ( hasField($result, 'message') && $result->message ) {
                $messages = Message_model::convertMessages($result->message);
            }
            return $messages;
        } catch ( Exception $ex ) {
            log_message(ERROR, 'Last messages per request failed! '.$ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
//    /**
//     * 
//     * @param long $adId
//     * @return array of Message_model
//     * @throws Exception
//     */
//    public function getMessagesByAd($adId) {
//        try {
//            $messageService = new SoapClient(MESSAGE_SERVICE_WSDL, getSoapOptions(loadToken()));
//            $result = $messageService->getMessagesByAd(array("adId" => $adId));
//            
//            $messages = array();
//            if ( hasField($result, 'message') && $result->message ) {
//                $messages = Message_model::convertMessages($result->message);
//            }
//            return $messages;
//        } catch ( Exception $ex ) {
//            log_message(ERROR, 'Messages request for ad (adId: '.$adId.') failed! '.$ex->faultstring);
//            throw new Exception($ex->faultstring);
//        }
//    }
    
    /**
     * 
     * @param long $requestId
     * @return array of Message_model
     * @throws Exception
     */
    public function getMessagesByRequest($requestId) {
        try {
            $messageService = new SoapClient(MESSAGE_SERVICE_WSDL, getSoapOptions(loadToken()));
            $result = $messageService->getMessagesByRequest(array("requestId" => $requestId));
            
            $messages = array();
            if ( hasField($result, 'message') && $result->message ) {
                $messages = Message_model::convertMessages($result->message);
            }
            return $messages;
        } catch ( Exception $ex ) {
            log_message(ERROR, 'Getting messages for request (requestId: '.$requestId.') failed! '.$ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    /**
     * 
     * @param long $user1Id
     * @param long $user2Id
     * @return array of Message_model
     * @throws Exception
     */
    public function getMessagesByUsers($user1Id, $user2Id) {
        try {
            $messageService = new SoapClient(MESSAGE_SERVICE_WSDL, getSoapOptions(loadToken()));
            $result = $messageService->getMessagesByUsers(array(
                "user1Id" => $user1Id,
                "user2Id" => $user2Id
            ));
            
            $messages = array();
            if ( hasField($result, 'message') && $result->message ) {
                $messages = Message_model::convertMessages($result->message);
            }
            return $messages;
        } catch ( Exception $ex ) {
            log_message(ERROR, 'Messages request for users (user1Id: '.$user1Id.', user2Id: '.$user2Id.') failed! '.$ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    /**
     * 
     * @param long $adId
     * @param long $user1Id
     * @param long $user2Id
     * @return array of Message_model
     * @throws Exception
     */
    public function getMessagesByAdAndUsers($adId, $user1Id, $user2Id) {
        try {
            $messageService = new SoapClient(MESSAGE_SERVICE_WSDL, getSoapOptions(loadToken()));
            $result = $messageService->getMessagesByAdAndUsers(array(
                "adId" => $adId,
                "user1Id" => $user1Id,
                "user2Id" => $user2Id
            ));
            
            $messages = array();
            if ( hasField($result, 'message') && $result->message ) {
                $messages = Message_model::convertMessages($result->message);
            }
            return $messages;
        } catch ( Exception $ex ) {
            log_message(ERROR, 'Messages request for ad (adId: '.$adId.') and users (user1Id: '.$user1Id.', user2Id: '.$user2Id.') failed! '.$ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    /**
     * 
     * @param Message_model $message
     * @return long message id
     * @throws Exception
     */
    public function sendMessage($message) {
        try {
            $messageService = new SoapClient(MESSAGE_SERVICE_WSDL, getSoapOptions(loadToken()));
            $result = $messageService->sendMessage(array("message" => $message));
            return $result->messageId;
        } catch ( Exception $ex ) {
            log_message(ERROR, 'Adding message failed! '.$ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }

    /**
     * 
     * @param Message_model $message
     * @throws Exception
     */
    public function updateMessage($message) {
        try {
            $messageService = new SoapClient(MESSAGE_SERVICE_WSDL, getSoapOptions(loadToken()));
            $messageService->updateMessage(array("message" => $message));
        } catch ( Exception $ex ) {
            log_message(ERROR, 'Update message failed! '.$ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    /**
     * 
     * @return array of Message_model
     * @throws Exception
     */
    public function getAllMessages() {
        try {
            $messageService = new SoapClient(MESSAGE_SERVICE_WSDL, getSoapOptions(loadToken()));
            $result = $messageService->getCommentsByAd();
            
            $messages = array();
            if ( hasField($result, 'message') && $result->message ) {
                $messages = Message_model::convertMessages($result->message);
            }
            return $messages;
        } catch ( Exception $ex ) {
            log_message(ERROR, 'Getting all messages failed! '.$ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    /**
     * 
     * @param long $messageId
     * @throws Exception
     */
    public function hideMessage($messageId) {
        try {
            $messageService = new SoapClient(MESSAGE_SERVICE_WSDL, getSoapOptions(loadToken()));
            $messageService->hideMessage(array("messageId" => $messageId));
        } catch ( Exception $ex ) {
            log_message(ERROR, 'Hide message (messageId: ' . $messageId . ') failed! '.$ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }
    
    /**
     * 
     * @param long $messageId
     * @throws Exception
     */
    public function deleteMessage($messageId) {
        try {
            $messageService = new SoapClient(MESSAGE_SERVICE_WSDL, getSoapOptions(loadToken()));
            $messageService->deleteMessage(array("messageId" => $messageId));
        } catch ( Exception $ex ) {
            log_message(ERROR, 'Delete message (messageId: ' . $messageId . ') failed! '.$ex->faultstring);
            throw new Exception($ex->faultstring);
        }
    }

    //*********
    //* share *
    //*********
    
    //...
}

File: trunk/venefica-webapp/src/main/java/com/venefica/service/MessageService.java
package com.venefica.service;

import com.venefica.service.dto.CommentDto;
import com.venefica.service.dto.MessageDto;
import com.venefica.service.fault.AdNotFoundException;
import com.venefica.service.fault.AuthorizationException;
import com.venefica.service.fault.CommentNotFoundException;
import com.venefica.service.fault.CommentValidationException;
import com.venefica.service.fault.MessageNotFoundException;
import com.venefica.service.fault.MessageValidationException;
import com.venefica.service.fault.RequestNotFoundException;
import com.venefica.service.fault.UserNotFoundException;
import java.util.List;
import javax.jws.WebMethod;
import javax.jws.WebParam;
import javax.jws.WebResult;
import javax.jws.WebService;
import javax.jws.soap.SOAPBinding;
import javax.jws.soap.SOAPBinding.ParameterStyle;
import javax.validation.constraints.NotNull;

/**
 * Allows to add comments to ads and sends messages directly to users.
 *
 * @author Sviatoslav Grebenchukov
 */
@WebService(name = "MessageService", targetNamespace = Namespace.SERVICE)
@SOAPBinding(parameterStyle = ParameterStyle.WRAPPED)
public interface MessageService {

    //**************
    //* commenting *
    //**************
    
    /**
     * Adds comment (question, answer, etc.) to the ad.
     *
     * @param adId the id of the ad
     * @param messageDto the comment to add
     * @return id of the created message
     * @throws CommentValidationException if commentDto doesn't contain required
     * fields
     */
    @WebMethod(operationName = "AddCommentToAd")
    @WebResult(name = "commentId")
    Long addCommentToAd(
            @WebParam(name = "adId") @NotNull Long adId,
            @WebParam(name = "comment") @NotNull CommentDto commentDto)
            throws AdNotFoundException, CommentValidationException;

    /**
     * Updates the comment in the database.
     *
     * @param commentDto updated comment
     * @throws CommentValidationException if commentDto doesn't contain required
     * fields
     */
    @WebMethod(operationName = "UpdateComment")
    void updateComment(@WebParam(name = "comment") @NotNull CommentDto commentDto)
            throws CommentNotFoundException, CommentValidationException;

    /**
     * Returns a list of comments by the ad.
     *
     * @param id if of the ad
     * @param lastCommentId the id of the last retrieved comment of -1 to return
     * all comments
     * @param numComments the maximum number of comments to return
     * @return list of comments
     * @throws AdNotFoundException if an ad with the specified id not found
     */
    @WebMethod(operationName = "GetCommentsByAd")
    @WebResult(name = "comment")
    List<CommentDto> getCommentsByAd(
            @WebParam(name = "adId") @NotNull Long adId,
            @WebParam(name = "lastCommentId") Long lastCommentId,
            @WebParam(name = "numComments") int numComments) throws AdNotFoundException;

    
    
    //*************
    //* messaging *
    //*************
    
    /**
     * Returns the last message for every request.
     * 
     * @return 
     */
    @WebMethod(operationName = "GetLastMessagePerRequest")
    @WebResult(name = "message")
    List<MessageDto> getLastMessagePerRequest();
    
//    /**
//     * Returns all messages associated with the given ad.
//     * 
//     * @param adId
//     * @return
//     * @throws AdNotFoundException 
//     */
//    @WebMethod(operationName = "GetMessagesByAd")
//    @WebResult(name = "message")
//    List<MessageDto> getMessagesByAd(@WebParam(name = "adId") @NotNull Long adId) throws AdNotFoundException;
    
    /**
     * Returns all messages associated with the given request.
     * 
     * @param requestId
     * @return
     * @throws RequestNotFoundException 
     */
    @WebMethod(operationName = "GetMessagesByRequest")
    @WebResult(name = "message")
    List<MessageDto> getMessagesByRequest(@WebParam(name = "requestId") @NotNull Long requestId) throws RequestNotFoundException, AuthorizationException;
    
    /**
     * Returns all the messages between the two given users.
     * 
     * @param user1Id
     * @param user2Id
     * @return
     * @throws UserNotFoundException if one of the users does not exists
     */
    @WebMethod(operationName = "GetMessagesByUsers")
    @WebResult(name = "message")
    List<MessageDto> getMessagesByUsers(
            @WebParam(name = "user1Id") @NotNull Long user1Id,
            @WebParam(name = "user2Id") @NotNull Long user2Id)
            throws UserNotFoundException;
    
    /**
     * Returns all the messages between the two given users for the specified ad.
     * 
     * @param adId
     * @param user1Id
     * @param user2Id
     * @return
     * @throws AdNotFoundException if the given ad does not exists
     * @throws UserNotFoundException if one of the users does not exists
     */
    @WebMethod(operationName = "GetMessagesByAdAndUsers")
    @WebResult(name = "message")
    List<MessageDto> getMessagesByAdAndUsers(
            @WebParam(name = "adId") @NotNull Long adId,
            @WebParam(name = "user1Id") @NotNull Long user1Id,
            @WebParam(name = "user2Id") @NotNull Long user2Id)
            throws AdNotFoundException, UserNotFoundException;
    
    /**
     * Sends the message to the specified user.
     *
     * @param message message dto object
     * @return id of the created message
     * @throws UserNotFoundException if a user with the specified name not found
     * @throws MessageValidationException if messageDto contains invalid fields.
     */
    @WebMethod(operationName = "SendMessage")
    @WebResult(name = "messageId")
    Long sendMessage(@WebParam(name = "message") @NotNull MessageDto messageDto)
            throws UserNotFoundException, RequestNotFoundException, MessageValidationException;

    /**
     * Updates message sent by the current user.
     *
     * @param messageDto updated message
     * @throws MessageNotFoundException if a message with the specified id not
     * found
     * @throws AuthorizationException if not owner of the message tries to
     * update it
     * @throws MessageValidationException if updated message doesn't contain
     * required fields
     */
    @WebMethod(operationName = "UpdateMessage")
    void updateMessage(@WebParam(name = "message") @NotNull MessageDto messageDto)
            throws MessageNotFoundException, AuthorizationException, MessageValidationException;

    /**
     * Get all messages sent to the current user or received by him.
     *
     * @return list of messages
     */
    @WebMethod(operationName = "GetAllMessages")
    @WebResult(name = "message")
    List<MessageDto> getAllMessages();
    
    /**
     * Not published via WS.
     * 
     * @param userId
     * @return
     * @throws UserNotFoundException 
     */
    int getUnreadMessagesSize(Long userId) throws UserNotFoundException;
    
    /**
     * Marks the message as hidden one.
     *
     * @param messageId id of the message
     * @throws MessageNotFoundException if a message with the specified id not
     * found
     * @throws AuthorizationException if we are neither recipient nor sender of
     * the message
     */
    @WebMethod(operationName = "HideMessage")
    void hideMessage(@WebParam(name = "messageId") @NotNull Long messageId) throws MessageNotFoundException,
            AuthorizationException;

    /**
     * Removes the message completely from the database.
     *
     * @param messageId id of the message
     * @throws MessageNotFoundException if a message with the specified id not
     * found
     * @throws AuthorizationException if we are neither recipient nor sender of
     * the message
     */
    @WebMethod(operationName = "DeleteMessage")
    void deleteMessage(@WebParam(name = "messageId") @NotNull Long messageId)
            throws MessageNotFoundException, AuthorizationException;
    
    
    
    //*********
    //* share *
    //*********
    
    /**
     * Places the message on the walls of the connected social networks.
     *
     * @param message text of the message
     */
    @WebMethod(operationName = "ShareOnSocialNetworks")
    void shareOnSocialNetworks(@WebParam(name = "message") @NotNull String message);
}
